### DISCUSSION TITLE
New Bottom Sheet Control in MCT

### DISCUSSION THREAD

**Discussion by eramrit78** [CommunityToolkit/Maui#1893] [Polls]
Now a days in most apps Android and IOS bottom sheet used in lot of screens. You can VOT it to add in MCT.

Thanks

**Comment by bijington** [comment]
How would this apply to macOS and Windows?

**Reply by eramrit78** [reply]
We can use a Bottom Bar for MAC and Windows like This:

![image](https://github.com/CommunityToolkit/Maui/assets/48124576/242783f8-54d9-4908-b2a0-2568cbbab7f8)

**Comment by bijington** [comment]
Is this a feature that you are willing to contribute into the toolkit?

**Reply by eramrit78** [reply]
No, The community will do it if possible.

**Comment by VladislavAntonyuk** [comment]
https://vladislavantonyuk.github.io/articles/Creating-a-bottom-sheet-using-.NET-MAUI

**Comment by vhugogarcia** [comment]
This library is widely used by a lot of developers, including myself. I highly recommend it since it used native implementation on each platform: https://github.com/the49ltd/The49.Maui.BottomSheet even our friend Gerald has mentioned it in 2 videos.
- https://www.youtube.com/watch?v=JJUm58avADo
- https://www.youtube.com/watch?v=bdKWnddRDY0

I think we should **not** implement a Bottom Sheet in MCT since there are already out there some solid options.

**Reply by peruchali** [reply]
> This library is widely used by a lot of developers, including myself. I highly recommend it since it used native implementation on each platform: https://github.com/the49ltd/The49.Maui.BottomSheet even our friend Gerald has mentioned it in 2 videos.
> 
> * https://www.youtube.com/watch?v=JJUm58avADo
> * https://www.youtube.com/watch?v=bdKWnddRDY0
> 
> I think we should **not** implement a Bottom Sheet in MCT since there are already out there some solid options.

These bottom sheet implementations don't work very well if your app has a flyout (it shows up above flyouts which is weird), or if you want it to show above tabbar/navigation bar (like Google Maps). It'd be awesome to have a bottomsheet implementation in MCT that actually works and is customizable.

**Reply by peruchali** [reply]
> This library is widely used by a lot of developers, including myself. I highly recommend it since it used native implementation on each platform: https://github.com/the49ltd/The49.Maui.BottomSheet even our friend Gerald has mentioned it in 2 videos.
> 
> * https://www.youtube.com/watch?v=JJUm58avADo
> * https://www.youtube.com/watch?v=bdKWnddRDY0
> 
> I think we should **not** implement a Bottom Sheet in MCT since there are already out there some solid options.

The49 BottomSheet repo has not been updated for a year, it is busted with .NET 9, which is the standard issue with unsupported 3rd party options. BottomSheets are so ubiquitous with every Android app (especially anything that uses a Map) that it should have a stable implementation for .NET MAUI. Request folks to reconsider.

**Reply by BillyMartin1964** [reply]
I really like the 49 bottom sheet because of the detents. Is there another bottom sheet out there that does that?

**Reply by vhugogarcia** [reply]
> I really like the 49 bottom sheet because of the detents. Is there another bottom sheet out there that does that?

Yes, I tried this recently and it is updated very often: https://github.com/lucacivale/Maui.BottomSheet

**Comment by stephenquan** [comment]
You can create a simple BottomSheet appearance with minimal customization of `Popup`. The key elements are:

 - Popup.HorizontalOptions="Fill"
 - Popup.VerticalOptions="End"
 - Popup.Color-"Transparent"
 - Border.StrokeShape="RoundRectangle 10,10,0,0"

The rest is customizing the appearance to suit your requirements

```xaml
<toolkit:Popup
    x:Class="MauiVariablePopup.ResizePopup"
    xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
    xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
    HorizontalOptions="Fill"
    VerticalOptions="End"
    Color="Transparent">
    <Border BackgroundColor="White" MaximumWidthRequest="600" StrokeShape="RoundRectangle 10,10,0,0">
        <VerticalStackLayout>
            <Label Margin="10,10,10,10" HorizontalOptions="Center" Text="Bottom Sheet Title" />
            <Border Margin="0,5,0,5" />
            <VerticalStackLayout Margin="20,0,20,0">
                <RadioButton Content="Yes" />
                <RadioButton Content="No" />
            </VerticalStackLayout>
            <Border Margin="0,5,0,5" />
            <Button Margin="20,10,20,10" HorizontalOptions="End" Text="Vote" />
        </VerticalStackLayout>
    </Border>
</toolkit:Popup>
```

**Reply by peruchali** [reply]
How does this work for non-modal bottomsheets? There's lot of functionality that a bottomsheet provides like detents, etc.

**Comment by CliffAgius** [comment]
Closed as answered, if closed in error please let us know to re-open.

**Reply by peruchali** [reply]
I don't think this is answered. There are workarounds, but those fall short of an actual bottomsheet control.

**Reply by bijington** [reply]
I think the answer is we aren't going to be implementing this in the toolkit. There aren't enough votes on the request and from what I can tell nobody is willing to implement it.

**Comment by stephenquan** [comment]
@peruchali non-modal behavior can be configured by setting CanBeDismissedByTappingOutsideOfPopup to False. A non-modal BottomSheet/BottomBar can also be achieved with DockLayout.

@peruchali as to the detents one way that you may achieve this is if you use an invisible rotated `Slider` to intercept the user swipe gesture and I also use `Slider` to drive `TranslationY` of both the bottom sheet grip and the bottom sheet content. Because It's in a ContentView the user can supply their own content via the ContentPresenter.

```xaml
<!-- BottomSheetView.xaml -->
<?xml version="1.0" encoding="utf-8" ?>
<ContentView
    x:Class="Maui.StackOverflow.BottomSheetView"
    xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
    xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit">
    <ContentView.Resources>
        <toolkit:MathExpressionConverter x:Key="Math" />
        <toolkit:MultiMathExpressionConverter x:Key="MultiMath" />
    </ContentView.Resources>
    <ContentView.ControlTemplate>
        <ControlTemplate>
            <Grid x:Name="bs">
                <Border
                    x:Name="grip"
                    Padding="1"
                    BackgroundColor="{AppThemeBinding Light=White, Dark=Black}"
                    HeightRequest="40"
                    Stroke="Gray"
                    StrokeShape="RoundRectangle 20, 20, 0, 0"
                    StrokeThickness="{TemplateBinding BorderWidth}"
                    TranslationY="{MultiBinding {Binding Value, Source={Reference slider}},
                                                {Binding Height, Source={Reference bs}},
                                                Converter={StaticResource MultiMath},
                                                ConverterParameter='x - x1 + 2'}"
                    VerticalOptions="End">
                    <Rectangle
                        BackgroundColor="Gray"
                        HeightRequest="4"
                        WidthRequest="40" />
                </Border>
                <Slider
                    x:Name="slider"
                    HeightRequest="{Binding Width, Source={Reference bs}}"
                    Maximum="{Binding Height, Source={Reference bs}}"
                    Minimum="{Binding Height, Source={Reference grip}}"
                    Opacity="0"
                    Rotation="90"
                    WidthRequest="{Binding Height, Source={Reference bs}}"
                    ZIndex="1"
                    Value="{Binding Height, Source={Reference bs}, Converter={StaticResource Math}, ConverterParameter='x/2'}" />
                <Border
                    x:Name="content"
                    Padding="1"
                    BackgroundColor="{AppThemeBinding Light=White, Dark=Black}"
                    HeightRequest="{MultiBinding {Binding Height, Source={Reference bs}},
                                                 {Binding Value, Source={Reference slider}},
                                                 Converter={StaticResource MultiMath},
                                                 ConverterParameter='x - x1 + 2'}"
                    HorizontalOptions="Fill"
                    TranslationY="{Binding Value, Source={Reference slider}}"
                    VerticalOptions="Start"
                    ZIndex="2">
                    <ContentPresenter />
                </Border>
            </Grid>
        </ControlTemplate>
    </ContentView.ControlTemplate>
</ContentView>
```

**Comment by BillyMartin1964** [comment]
https://github.com/the49ltd/The49.Maui.BottomSheet

The49.Maui.BottomSheet  is the best I've seen so far, but it isn't being maintained and it is crashing in .net 9. (in Android, at least)

This would be a great starting point for Community toolkit.
